File: java/src/main/java/com/dataart/javathrifthelloworld/DummyServiceImpl.java
package com.dataart.javathrifthelloworld;

import org.apache.thrift.TException;

import java.util.HashMap;
import java.util.Map;

public class DummyServiceImpl implements DummyService.Iface {
    private final Map<String, Person> people = new HashMap<String, Person>();

    @Override
    public void putPerson(Person person) throws TException {
        people.put(person.getId(), person);
    }

    @Override
    public Person getPerson(String id) throws PersonNotFoundException, TException {
        Person person = people.get(id);
        if(person == null) {
            String message = String.format("No person with id %s", id);
            throw new PersonNotFoundException(id, message);
        }

        return person;
    }

    @Override
    public int getPersonCount() throws TException {
        return people.size();
    }
}

File: cpp/src/server/DummyServiceImpl.cpp
#include "DummyServiceImpl.h"
#include <map>

DummyServiceHandler::DummyServiceHandler()
{
}

DummyServiceHandler::~DummyServiceHandler()
{
}

void DummyServiceHandler::putPerson(const Person& person)
{
	printf("putPerson\n");
	people[person.id] = person;
}

void DummyServiceHandler::getPerson(Person& _return, const std::string& id)
{
	printf("getPerson\n");
	std::map<std::string, Person>::iterator it  = people.find(id);
	if (it == people.end())
	{
		PersonNotFoundException e;
		e.requestedId = id;
		e.errorMessage = "No person with id " + id;
		throw e;
	}
	else
		_return = it->second;
}

int32_t DummyServiceHandler::getPersonCount()
{
	printf("getPersonCount\n");
	return people.size();
}

